drivers/interrupt_controller/intc_nxp_irqsteer.c

```diff
@@ -0,0 +1,522 @@
+	}
+
+	return idx;
+}
+
+void z_soc_irq_enable_disable(uint32_t irq, bool enable)
+{
+	uint32_t parent_irq;
+	int i, system_irq, level2_irq;
+	const struct irqsteer_config *cfg;
+
+	if (irq_get_level(irq) == 1) {
+		/* LEVEL 1 interrupts are DSP direct */
+		z_xtensa_irq_enable(XTENSA_IRQ_NUMBER(irq));
+		return;
+	}
+
+	parent_irq = irq_parent_level_2(irq);
+	level2_irq = irq_from_level_2(irq);
+
+	/* find dispatcher responsible for this interrupt */
+	for (i = 0; i < ARRAY_SIZE(dispatchers); i++) {
+		if (dispatchers[i].irq == parent_irq) {
+			cfg = dispatchers[i].dev->config;
+
+			system_irq = from_zephyr_irq(cfg->regmap_phys,
+						     level2_irq,
+						     dispatchers[i].master_index);
+
+			if (enable) {
+				IRQSTEER_EnableInterrupt((IRQSTEER_Type *)cfg->regmap_phys,
+							 system_irq);
+			} else {
+				IRQSTEER_DisableInterrupt((IRQSTEER_Type *)cfg->regmap_phys,
+							  system_irq);
+			}
+
+			break;
+		}
+	}
+
+	__ASSERT(false, "INTID %d doesn't belong to any dispatcher", irq);
+}
+
+void z_soc_irq_enable(uint32_t irq)
+{
+	z_soc_irq_enable_disable(irq, true);
+}
+
+void z_soc_irq_disable(uint32_t irq)
+{
+	z_soc_irq_enable_disable(irq, false);
+}
+
+int z_soc_irq_is_enabled(unsigned int irq)
+{
+	uint32_t parent_irq;
+	int i, system_irq, level2_irq;
+	const struct irqsteer_config *cfg;
+
+	if (irq_get_level(irq) == 1) {
+		/* LEVEL 1 interrupts are DSP direct */
+		return z_xtensa_irq_is_enabled(XTENSA_IRQ_NUMBER(irq));
+	}
+
+	parent_irq = irq_parent_level_2(irq);
+	level2_irq = irq_from_level_2(irq);
+
+	/* find dispatcher responsible for this interrupt */
+	for (i = 0; i < ARRAY_SIZE(dispatchers); i++) {
+		if (dispatchers[i].irq == parent_irq) {
+			cfg = dispatchers[i].dev->config;
+
+			system_irq = from_zephyr_irq(cfg->regmap_phys,
+						     level2_irq,
+						     dispatchers[i].master_index);
+
+			return IRQSTEER_InterruptIsEnabled((IRQSTEER_Type *)cfg->regmap_phys,
+							   system_irq);
+		}
+	}
+
+	__ASSERT(false, "INTID %d doesn't belong to any dispatcher", irq);
+
+	return false;
+}
+
+
+static void irqsteer_isr_dispatcher(const void *data)
+{
+	struct irqsteer_dispatcher *dispatcher;
+	const struct irqsteer_config *cfg;
+	uint32_t irq_num, table_idx;
+	int i, counter;
+	int system_irq, zephyr_irq;
+
+	dispatcher = (struct irqsteer_dispatcher *)data;
+	cfg = dispatcher->dev->config;
+	irq_num = IRQSTEER_GetMasterIrqCount((IRQSTEER_Type *)cfg->regmap_phys,
+					     dispatcher->master_index);
+	i = 0;
+	counter = 0;
+
+	while (true) {
+		/* which system interrupt is responsible for triggering
+		 * dispatcher's interrupt?
+		 */
+		system_irq = IRQSTEER_GetMasterNextInterrupt((IRQSTEER_Type *)cfg->regmap_phys,
+							     dispatcher->master_index);
+		if (system_irq == NotAvail_IRQn)
+			break;
+
+		__ASSERT(system_irq >= 0,
+			 "got negative value for system IRQ");
+
+		if (i == irq_num) {
+			i = 0;
+			counter++;
+		}
+
+		/* TODO: what if asserts are disabled? Do we just let the
```

Comment:
Why don't you check the `counter` in an `if` and if it exceeds the threshold `break` and print an error message?

Reply:
Mm, won't the ISR still get called anyways in that case? The point of the `ASSERT()` was to break your system whenever you get an interrupt that keeps getting asserted.

I'm thinking about changing the `ASSERT()` to a `z_irq_spurious(NULL)` call which will for sure break your system. Technically, you weren't able to fully handle your interrupt so I guess I can make a case for calling `z_irq_spurious(NULL)` for other than just breaking the system. Let me know what you think

Reply:
The problem here is actually that `IRQSTEER_GetMasterNextInterrupt` always  returns the first active interrupt - this is, in fact, get first interrupt. So, if, let's say, interrupt 1 is triggered endlessly you always handle this interrupt (or, a maximum of `counter` times), never get to the next one from `CHn_STATUS`.

IMO, a proper fix will be to get all active interrupts (have a `IRQSTEER_GetMasterActiveInterrupts()`)  outside of `while()` and in `while()` handle each of these interrupts. 
In case another peripheral interrupt is trigger while you handle these, you'll have the `irqsteer_isr_dispatcher()` called again.

So, no need for a threshold.

Reply:
The possible issue that I see with this approach is that you'll no longer be able to distinguish between the case where `irqsteer_isr_dispatcher()` is called under normal circumstances and the case where it's called when you have a "rogue" interrupt. The point of the threshold is to try to detect the anomalies and at least print an error message so that the user knows it's an interrupt-related issue. Of course, this is really dependent on the scenario (e.g: you can have an interrupt getting triggered abnormally fast for some reason but not fast enough that it will get discovered by the threshold).

I guess another potential issue with the non-`IRQSTEER_GetMasterNextInterrupt()` approach is that you're going to have to add some extra complexity in the driver since a master may have 2 groups of 32 interrupts (master1 and master2, i.MX8MP) aggregated by it instead of just 1 (master0, i.MX8MP). At this point I'm wondering if the non-`IRQSTEER_GetMasterNextInterrupt()` approach is really worth it?

Reply:
> The possible issue that I see with this approach is that you'll no longer be able to distinguish between the case where `irqsteer_isr_dispatcher()` is called under normal circumstances and the case where it's called when you have a "rogue" interrupt. The point of the threshold is to try to detect the anomalies and at least print an error message so that the user knows it's an interrupt-related issue. Of course, this is really dependent on the scenario (e.g: you can have an interrupt getting triggered abnormally fast for some reason but not fast enough that it will get discovered by the threshold). **EDIT**: At this point this is just a theoretical talk on my side as I haven't really explored the case in which an interrupt gets triggered endlessly.
> 
A "rogue" interrupt should not be handle in irq_steer, this is a peripheral issue - the irq is not cleared properly or something else.

> I guess another potential issue with the non-`IRQSTEER_GetMasterNextInterrupt()` approach is that you're going to have to add some extra complexity in the driver since a master may have 2 groups of 32 interrupts (master1 and master2, i.MX8MP) aggregated by it instead of just 1 (master0, i.MX8MP). At this point I'm wondering if the non-`IRQSTEER_GetMasterNextInterrupt()` approach is really worth it?

First, `IRQSTEER_GetMasterNextInterrupt` is definitely not NextInterrupt, is FirstInterrupt.
Second, the approach of getting all active interrupt and handle them in `while()` I believe is much simpler.
You can have an uint64_t and each bit represents the active interrupt.
In case of 2 masters each with 32 interrupts, you can have first 32 bits for one master, and the rest for the other one.

Reply:
In the current design I see the following problems:
1. if a peripheral has an issue and triggers interrupts you never move to the next one - you handle that one for 5 times and exit;
2. iterate over 64 (or in case of master 0, 32) interrupts for 5 times and might handle only 1 interrupt - see 1.

With the proposed solution you'll iterate only over the activate interrupts which might be max 64.

Reply:
ACK, will fix this